src/transaction/log_recovery_redo_parallel.hpp

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright 2008 Search Solution Corporation
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#ifndef LOG_RECOVERY_REDO_HPP
+#define LOG_RECOVERY_REDO_HPP
+
+#include "log_compress.h"
+#include "log_recovery_redo.hpp"
+#include "thread_manager.hpp"
+#include "thread_worker_pool.hpp"
+
+#include <bitset>
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <set>
+
+namespace cublog
+{
+  /* contains a single vpid-backed log record entry to be applied on a page;
+   * this is a base class designed to allow handling in a container;
+   * internal implementation detail
+   *
+   * NOTE: an optimization would be to allow this to be a container of more than
+   *  one same-vpid log record entry to be applied in one go by the same thread
+   */
+  class redo_job_base
+  {
+    public:
+      redo_job_base (VPID a_vpid)
+	: vpid (a_vpid)
+      {
+	// the logic implemented below, in the queue and task, does allow for null-vpid type of
+	// entries to be executed - they are called "to be waited for" or "synched" operations;
+	// but, for now, do not allow them to be dispatched to be executed asynchronously
+	assert (!VPID_ISNULL (&vpid));
+      }
+      redo_job_base () = delete;
+      redo_job_base ( redo_job_base const &) = delete;
+      redo_job_base ( redo_job_base &&) = delete;
+
+      virtual ~redo_job_base() = default;
+
+      redo_job_base &operator = ( redo_job_base const &) = delete;
+      redo_job_base &operator = ( redo_job_base &&) = delete;
+
+      /* log entries come in more than one flavor:
+       *  - pertain to a certain vpid - aka: page update
+       *  - pertain to a certain VOLID - aka: volume extend, new page
+       *  - pertain to no VOLID - aka: database extend, new volume
+       */
+      const VPID &get_vpid() const
+      {
+	return vpid;
+      }
+
+      virtual int execute (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+			   LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support) = 0;
+
+      bool get_is_to_be_waited_for() const
+      {
+	return vpid.volid == NULL_VOLID || vpid.pageid == NULL_PAGEID;
+      }
+
+    private:
+      const VPID vpid;
+  };
+  using ux_redo_job_base = std::unique_ptr<redo_job_base>;
```

Comment:
Please don't use aliases for std::unique_ptr in header. Better to use the full name, or use aliases in smaller scopes (unit files or inside a class/function)

Reply:
Done. Kept `typedef`s only as part of the internal implementation.